exchange/non_bid_reason.go

```diff
@@ -1,29 +1,55 @@
 
 const (
-	NoBidUnknownError                      NonBidReason = 0 // No Bid - General
-	ResponseRejectedGeneral                NonBidReason = 300
-	ResponseRejectedBelowFloor             NonBidReason = 301 // Response Rejected - Below Floor
-	ResponseRejectedCategoryMappingInvalid NonBidReason = 303 // Response Rejected - Category Mapping Invalid
-	ResponseRejectedBelowDealFloor         NonBidReason = 304 // Response Rejected - Bid was Below Deal Floor
-	ResponseRejectedCreativeSizeNotAllowed NonBidReason = 351 // Response Rejected - Invalid Creative (Size Not Allowed)
-	ResponseRejectedCreativeNotSecure      NonBidReason = 352 // Response Rejected - Invalid Creative (Not Secure)
+	ResponseRejectedGeneral                openrtb3.NoBidReason = 300
+	ResponseRejectedBelowFloor             openrtb3.NoBidReason = 301 // Response Rejected - Below Floor
+	ResponseRejectedCategoryMappingInvalid openrtb3.NoBidReason = 303 // Response Rejected - Category Mapping Invalid
+	ResponseRejectedBelowDealFloor         openrtb3.NoBidReason = 304 // Response Rejected - Bid was Below Deal Floor
+	ResponseRejectedCreativeSizeNotAllowed openrtb3.NoBidReason = 351 // Response Rejected - Invalid Creative (Size Not Allowed)
+	ResponseRejectedCreativeNotSecure      openrtb3.NoBidReason = 352 // Response Rejected - Invalid Creative (Not Secure)
+	ErrorTimeout                           openrtb3.NoBidReason = 101 // Error - Timeout
+	ErrorGeneral                           openrtb3.NoBidReason = 100 // Error - General
+	ErrorBidderUnreachable                 openrtb3.NoBidReason = 103 // Error - Bidder Unreachable
 )
 
-// Ptr returns pointer to own value.
-func (n NonBidReason) Ptr() *NonBidReason {
-	return &n
+func errorToNonBidReason(errorCode int) openrtb3.NoBidReason {
+	switch errorCode {
+	case errortypes.TimeoutErrorCode:
+		return ErrorTimeout
+	}
+	// return 0
+	return ErrorGeneral
 }
 
-// Val safely dereferences pointer, returning default value (NoBidUnknownError) for nil.
-func (n *NonBidReason) Val() NonBidReason {
-	if n == nil {
-		return NoBidUnknownError
+// httpInfoToNonBidReason determines NoBidReason code (NBR)
+// It will first try to resolve the NBR based on prebid's proprietary error code.
+// If proprietary error code not found then it will try to determine NBR using
+// system call level error code
+func httpInfoToNonBidReason(httpInfo *httpCallInfo) openrtb3.NoBidReason {
+	errorCode := errortypes.ReadCode(httpInfo.err)
+	nonBidReason := errorToNonBidReason(errorCode)
+	if nonBidReason == ErrorGeneral {
+		if isBidderUnreachableError(httpInfo) {
+			return ErrorBidderUnreachable
+		}
 	}
-	return *n
+	return nonBidReason
+}
+
+// isBidderUnreachableError checks if the error is due to connection refused or no such host
+func isBidderUnreachableError(httpInfo *httpCallInfo) bool {
+	return errors.Is(httpInfo.err, syscall.ECONNREFUSED) || strings.Contains(httpInfo.err.Error(), "no such host")
```

Comment:
It's dangerous to use string comparison on framework provided error messages. This affects us on unit tests and should not be used in production. You can replace this test with this approach:

```
var dnsErrr *net.DNSError
isNoSuchHost := errors.As(err, &dnsErr) && dnsErr.IsNotFound
```

Reply:
Done